src/cs50/sql.py

```diff
@@ -41,202 +44,316 @@ def __init__(self, url, **kwargs):
             return None
 
         # If user errror
         except sqlalchemy.exc.OperationalError as e:
-            self.logger.debug(termcolor.colored(log, "red"))
-            e = RuntimeError(self._parse(e))
+            self._logger.debug(termcolor.colored(statement, "red"))
+            e = RuntimeError(_parse_exception(e))
             e.__cause__ = None
             raise e
 
         # Return value
         else:
-            self.logger.debug(termcolor.colored(log, "green"))
+            self._logger.debug(termcolor.colored(statement, "green"))
             return ret
 
+    def _escape(self, value):
+        """
+        Escapes value using engine's conversion function.
+
+        https://docs.sqlalchemy.org/en/latest/core/type_api.html#sqlalchemy.types.TypeEngine.literal_processor
+        """
+
+        def __escape(value):
+
+            # bool
+            if type(value) is bool:
+                return sqlparse.sql.Token(
+                    sqlparse.tokens.Number,
+                    sqlalchemy.types.Boolean().literal_processor(self.engine.dialect)(value))
+
+            # datetime.date
+            elif type(value) is datetime.date:
+                return sqlparse.sql.Token(
+                    sqlparse.tokens.String,
+                    sqlalchemy.types.String().literal_processor(self.engine.dialect)(value.strftime("%Y-%m-%d")))
+
+            # datetime.datetime
+            elif type(value) is datetime.datetime:
+                return sqlparse.sql.Token(
+                    sqlparse.tokens.String,
+                    sqlalchemy.types.String().literal_processor(self.engine.dialect)(value.strftime("%Y-%m-%d %H:%M:%S")))
+
+            # datetime.time
+            elif type(value) is datetime.time:
+                return sqlparse.sql.Token(
+                    sqlparse.tokens.String,
+                    sqlalchemy.types.String().literal_processor(self.engine.dialect)(value.strftime("%H:%M:%S")))
+
+            # float
+            elif type(value) is float:
+                return sqlparse.sql.Token(
+                    sqlparse.tokens.Number,
+                    sqlalchemy.types.Float().literal_processor(self.engine.dialect)(value))
+
+            # int
+            elif type(value) is int:
+                return sqlparse.sql.Token(
+                    sqlparse.tokens.Number,
+                    sqlalchemy.types.Integer().literal_processor(self.engine.dialect)(value))
+
+            # str
+            elif type(value) is str:
+                return sqlparse.sql.Token(
+                    sqlparse.tokens.String,
+                    sqlalchemy.types.String().literal_processor(self.engine.dialect)(value))
+
+            # None
+            elif value is None:
+                return sqlparse.sql.Token(
+                    sqlparse.tokens.Keyword,
+                    sqlalchemy.types.NullType().literal_processor(self.engine.dialect)(value))
```

Comment:
Hm, to be honest, I'm not sure how to pass them at the moment, since we're reducing the whole statement to a `str` ourselves. And a query like

```
db.execute("SELECT * FROM Employee WHERE FirstName = ?", b'\x00')
```

currently errs with

```
ValueError: the query contains a null character
```

at this line of ours:

```
result = self.engine.execute(statement)
```

Know how/if we can represent binary data in strings in a cross-DBMS way? Or do we need to revert to binding parameters for those?

Reply:
PS, I started to add support with, e.g.:

```
            # bytearray, bytes
            elif type(value) in [bytearray, bytes]:
                return sqlparse.sql.Token(
                    sqlparse.tokens.Other,
                    sqlalchemy.types.LargeBinary().literal_processor(self.engine.dialect)(value))
```

Reply:
Moving this to a feature request, https://github.com/cs50/python-cs50/issues/76.